ckanext/dcat/plugins/__init__.py

```diff
@@ -132,6 +135,30 @@ def set_titles(object_dict):
 
         return data_dict
 
+    def before_dataset_index(self, dataset_dict):
+        schema = None
+        try:
+            schema_show = p.toolkit.get_action("scheming_dataset_schema_show")
+            try:
+                schema = schema_show({}, {"type": dataset_dict["type"]})
+            except p.toolkit.ObjectNotFound:
+                pass
+        except KeyError:
+            pass
+
+        if schema:
+            for field in schema['dataset_fields']:
+                if field['field_name'] in dataset_dict and 'repeating_subfields' in field:
+                    for index, item in enumerate(dataset_dict[field['field_name']]):
+                        for key in item:
+                            # Index a flattened version
+                            new_key = f'{field["field_name"]}_{index}_{key}'
```

Comment:
IIUC this would let us ask solr for things like 'datasets with the 6th name field in contacts equal to "frank"', but not 'datasets containing any contact named "frank"'. If we the same keys from all subfields into a single field like
```
new_key = f'{field["field_name"]}__{key}'
```
then we could, right? We're not doing dynamic solr schemas so we'd have to combine everything as text fields but it should work for text searches.

Also nothing here is specific to dcat so shouldn't it be a scheming PR?

Reply:
Right, given this:

![Screenshot 2024-05-23 at 13-26-38 Test dcat 1 - Dataset - CKAN](https://github.com/ckan/ckanext-dcat/assets/200230/d08506cb-ef8a-45c2-a51b-300983a67b22)

With your suggestion you need to run a query like `q=contact__name:*Racoon*` to get a hit, with the original it would be impossible to get this hit without knowing the index, so that's obviously better. 
But I think users would expect to find these results in a free text search as well. For this, the subfields need to be indexed in an `extras_*` field, as these are copied to the [catch-all `text`](https://github.com/ckan/ckan/blob/5334babe3888a239dea15a17c8af8e46d1c3f346/ckan/config/solr/schema.xml#L186) field. So this would allow just to do `q=Racoon` and get a result back, so maybe indexing 

```
new_key = f'extras_{field["field_name"]}__{key}'
```
with the combined key values is the better approach.

> Also nothing here is specific to dcat so shouldn't it be a scheming PR?

Sure, I was just getting things working in this extension. Do you want to replace the logic in [https://github.com/ckan/ckanext-scheming/blob/9a33dcd330211468ea0bed85391277e2e596f388/ckanext/scheming/plugins.py#L477](https://github.com/ckan/ckanext-scheming/blob/9a33dcd330211468ea0bed85391277e2e596f388/ckanext/scheming/plugins.py#L477) with this one or create a separate plugin? I'll send a PR